### ISSUE TITLE
Regression? can't load package: package _/...

### ISSUE THREAD

**Issue by RobertBaruch** [microsoft/vscode-go#2630]
Not sure how to reproduce this, since this appears to be a new bug introduced some time in the past two versions ("Unpacking code (1.36.0-1562161087) over (1.34.0-1557957934) ...")

This is on Linux.

Whenever I save a file, I get an error at the top "package X" line:

```
{
	"resource": "/usr/local/home/robertbaruch/go/src/project/tools/x/x.go",
	"owner": "go",
	"severity": 8,
	"message": "can't load package: package _/usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud: cannot find package \"_/usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud\" in any of:\n\t/usr/lib/golang/src/_/usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud (from $GOROOT)\n\t/usr/local/home/robertbaruch/go/src/_/usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud (from $GOPATH)\n (... tons of errors like this...)
	"source": "go",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 13
}

```
I noticed that it's looking for packages starting with an underscore, which I don't understand.

**Comment by ramya-rao-a** [comment]
underscore is Go's way of saying that it is working with files that are not under the GOPATH.
Run the command `Go: Current GOPATH` and see if it gives you the expected GOPATH which in your case should be `/usr/local/home/robertbaruch/go`

**Comment by RobertBaruch** [comment]
The GOPATH is correct: `/usr/local/home/robertbaruch/go`.

**Comment by RobertBaruch** [comment]
Just to make it clear, `/usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud` does exist.

**Comment by RobertBaruch** [comment]
Sadly, I downgraded back to 1.34.0, and no change. Then I downgraded the Go extension to 0.11.0, no change. Downgrade to 0.10.2, no change. So now I have no idea what changed to cause this behavior when I chose to upgrade to 1.36.0.

**Comment by RobertBaruch** [comment]
Hmm, it might be something I did and forgot about, I changed the setting Go: Build On Save from package to workspace. Going back to package made the error stop happening, but I'd really like workspace to work.

**Comment by ramya-rao-a** [comment]
Are you able to consistently see the problem when you switch the build on save setting to `workspace` and then see it go away when set to `package`?

When this happens, you can find the exact `go build` command used and the directory from where the command is run in the Output channel. `View` -> `Output` -> `Go` from he dropdown on the top right corner of the output panel.

Compare this for the 2 scenarios, maybe that will shed some light on the problem.

Also are you using modules?

**Comment by RobertBaruch** [comment]
Still happening, still goes away for `package`.

Here's the go build command. It's very very long, so I cut it down a bit:

`/usr/lib/golang/bin/go build _/usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud _/usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud/asset/apiv1` (here follows many many more packages in `pkg/dep/sources`, followed by my own packages, such as `pkg/change pkg/config`)

`can't load package: package _/usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud: cannot find package "_/usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud" in any of:
	/usr/lib/golang/src/_/usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud (from $GOROOT)
	/usr/local/home/robertbaruch/go/src/_/usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud (from $GOPATH)`

I wonder if this is an incompatibility with `[dep](https://github.com/golang/dep)`? My directory structure is like this:

```
go
go/bin
go/pkg  <-- maintained by dep
go/pkg/dep
go/pkg/dep/sources
go/pkg/dep/sources/https---code.googlesource.code-gocloud (etc)
go/src
go/src/project
go/src/project/pkg
go/src/project/pkg/util
go/src/project/pkg/util/util.go
go/src/project/tools
go/src/project/tools/x
go/src/project/tools/x.go
go/src/project/vendor  <-- generated by dep
go/src/project/vendor/cloud.google.com (etc)
```

So maybe the problem is that the `go build` command line being built for the workspace setting includes `$GOPATH/pkg`, when it should only include stuff under `$GOPATH/src`?

**Comment by RobertBaruch** [comment]
Hmm, yeah, I'd say don't compile anything except files under src. (ref: https://golang.org/doc/code.html)

**Comment by ramya-rao-a** [comment]
Thats some god observations. Can you also share the `go build` command that gets run when you change the build on save setting to `package`?

The extension runs `go list -f ImportPath: {{.ImportPath}} FolderPath: {{.Dir}} ./...` in your workspace root to get the list of packages to use in the `go build ` command. This seems to be returning things under `$GOPATH/pkg` for you. Can you run this command in your workspace folder and confirm?

**Comment by RobertBaruch** [comment]
When set to `package`, the command is (after saving in util.go)

```
/usr/lib/golang/bin/go build -o /tmp/vscode-goHHoCXV/go-code-check project/pkg/util
```
`go list -f "ImportPath: {{.ImportPath}} FolderPath: {{.Dir}}" ./...` from my workspace root outputs:

```
ImportPath: _/usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud FolderPath: /usr/local/home/robertbaruch/go/pkg/dep/sources/https---code.googlesource.com-gocloud
(then follows hundreds of subpackages from there)
ImportPath: project/pkg/util FolderPath: /usr/local/home/robertbaruch/go/src/project/pkg/util
ImportPath: project/x FolderPath: /usr/local/home/robertbaruch/go/src/project/x
```

On the other hand, specifying the last arg as `./src/...` lists all the correct paths, and also excludes `go/pkg`, which I think is more correct, given that the only source that a workspace should pay attention to is in `src`. I think using `./...` from the workspace root, referenced in various posts, assumes you don't have anything other than `go/bin` and `go/src`, which is not the case if you use `dep`.

**Comment by ramya-rao-a** [comment]
Why wouldnt you open the `src` directory instead of its parent in VS Code? Then `./...` would work correctly right?

The problem here is that the Go extension is not aware of dep or any code structure for that matter. It places its trust in `go list` and expects the output of this command to list the packages that it needs to build

**Comment by RobertBaruch** [comment]
Sure, opening in `src` would work. But that could conflict with the `inferGopath` setting, which infers `GOPATH` from the workspace root. I'm assuming that this just sets `GOPATH` to the **vscode** workspace root? It's a little confusing because GOPATH is the **Go** workspace root, which includes `bin` and `src`. 

So now you have two choices: 

- open up vscode in the Go workspace root and have the buildOnSave workspace setting fail if you're using `dep`, or 
- open up vscode in the Go workspace's `src` directory and have the buildOnSave workspace setting work, but have `inferGopath` fail if you didn't set `GOPATH`.

Basically there is an inconsistency here.

**Comment by RobertBaruch** [comment]
(incidentally, opening vscode in the `src` directory and setting buildOnSave to workspace does work)

**Comment by ramya-rao-a** [comment]
Hey @RobertBaruch 

Apologies for the late follow up here

Do you still have problems with `inferGopath` setting when you open vscode in the `src` directory?

I checked the code around how we infer go path based on the folder that was opened as a root in vscode, and it looks like things should work as expected even when the `src` directory is opened as root.

**Comment by RobertBaruch** [comment]
Seems to be working ok now. I've also switched to the Go language server, that could have something to do with it? But in any case, I no longer have any issues!
